OrbitCore/SymbolHelper.cpp

```diff
@@ -137,3 +161,36 @@ ErrorMessageOr<ModuleSymbols> SymbolHelper::LoadUsingSymbolsPathFile(
     const std::string& module_path, const std::string& build_id) const {
   return FindSymbols(module_path, build_id, symbols_file_directories_);
 }
+
+ErrorMessageOr<std::string> SymbolHelper::FindDebugSymbolsFile(
+    const std::string& module_path, const std::string& build_id) const {
+  return FindSymbolsFile(module_path, collector_symbol_directories_, build_id);
+}
+
+ErrorMessageOr<ModuleSymbols> SymbolHelper::LoadSymbolsFromFile(
+    const std::string& file_path, const std::string& build_id) const {
+  ErrorMessageOr<std::unique_ptr<ElfFile>> elf_file_result =
+      ElfFile::Create(file_path);
+
+  if (!elf_file_result) {
+    return ErrorMessage(
+        absl::StrFormat("Failed to load debug symbols from \"%s\": %s",
+                        file_path, elf_file_result.error().message()));
+  }
+
+  const std::string& target_build_id = elf_file_result.value()->GetBuildId();
+  if (target_build_id != build_id) {
+    return ErrorMessage(
+        absl::StrFormat("Failed to load debug symbols from \"%s\": invalid "
+                        "build id \"%s\", expected: \"%s\"",
+                        file_path, target_build_id, build_id));
+  }
+
+  return elf_file_result.value()->LoadSymbols();
+}
+
+std::string SymbolHelper::GenerateCachedFileName(
+    const std::string& file_path) const {
+  auto file_name = absl::StrReplaceAll(file_path, {{"/", "_"}});
+  return Path::JoinPath({Path::GetCachePath(), file_name});
```

Comment:
Did you consider also involving the build id here?

Reply:
This will increase necessary size of cache directly - currently we have a situation when if a module is updated it will override exiting (old) cache file, which saves space and is most common scenario. It is possible that when jumping between 2 version it will require do re-download file every time but I think it is more rare use-case. Having said that I do not feel strongly about this. It might make sense to add build_id to the name to avoid extra downloads.

Reply:
I am going to leave it as is for now.